### ISSUE TITLE
Apply TomcatConnectorCustomizer and TomcatContextCustomizer beans automatically

### ISSUE THREAD

**Issue by philwebb** [spring-projects/spring-boot#15062]
It would be nice if you could register `TomcatConnectorCustomizer` and `TomcatContextCustomizer` implementations as beans and have them applied automatically.

**Comment by philwebb** [comment]
See also #15035

**Comment by govi20** [comment]
I would like to work on this issue.

**Comment by mbhave** [comment]
@govi20 Sure! Feel free to ask any questions you might have. We'll look forward to your Pull Request.

**Comment by wilkinsona** [comment]
We already have [a PR](https://github.com/spring-projects/spring-boot/pull/15006) from you, @govi20. Thank you for that. In the spirit of first-timers only can we please carry on working with you on that and leave this issue for someone who hasn’t contributed before?

**Comment by nishantraut** [comment]
Hi @wilkinsona can i work on this?

**Comment by wilkinsona** [comment]
@nishantraut Thank you for the offer, but you have [already contributed](
https://github.com/spring-projects/spring-boot/pull/14977) as well. This issue is intended especially for someone new.

**Comment by nishantraut** [comment]
Yes @wilkinsona that was documentation update. How can i identify and pickup small issues in code.

**Comment by wilkinsona** [comment]
@nishantraut Please look at the `ideal-for-contribution` label as recommended above.

**Comment by xoronet** [comment]
Hi, i would like to try it!

**Comment by govi20** [comment]
I am yet to make the contribution, as per Stephane my pull request does not cover all the requirements and issue is very complex. Anyways, will work on something else.

**Comment by sadath42** [comment]
Should i create a branch from the master or should i fork the master

**Comment by wilkinsona** [comment]
@sadath42 Thanks for offering to work on this but @xoronet is already doing so. Please keep an eye out for other `first-timers-only` issues in the future and we'll be happy to guide you through the process once you've claimed an issue.

**Comment by wilkinsona** [comment]
How's it going, @xoronet? If you have any questions, please feel free to ask them here and we'll do our best to help.

**Comment by ihrigb** [comment]
If @xoronet is no longer available to look into this, I could take a shot. Just let me know.

**Comment by xoronet** [comment]
Hi @wilkinsona, fine so far. I have forked, cloned and build the project. I also digged into the topic as well as into the code. Unfortunately I had to focus on a work related project in the last days. Furthermore I had to check if my employee allows me to sign the CLA. No problems from his side. I will go back to solving the issue on Monday. Is there a date until the issue should be finished?

**Comment by wilkinsona** [comment]
Great, thank you. The issue's currently in the 2.1.x milestone and 2.1.1 is scheduled for 27 November. It would be nice, although it's not essential, for the issue to be fixed before then.

**Comment by wilkinsona** [comment]
@Ksaboor You can fine other first-timer-only issues by clicking on the label. At the time of writing this is the only such open first-timers issue.

**Comment by xoronet** [comment]
Hi guys, I am terrible sorry but I am not able to work on the topic because of too much other tasks at work. I dont find enough time, so I think it would be the best to return the issue to the crowd. I am terrible sorry for wasting your time! @philwebb @wilkinsona @Razi007

**Comment by wilkinsona** [comment]
Thanks for letting us know, @xoronet.

@ihrigb I think you were next in line. Would you like to take this on one?

**Comment by ihrigb** [comment]
@wilkinsona, yes I will take care of it.

**Comment by wilkinsona** [comment]
Sorry, @ihrigb. I've just noticed that you've already made [two contributions](https://github.com/spring-projects/spring-boot/pulls?utf8=✓&q=is%3Apr+author%3Aihrigb) to the project. Thank you again for those. As a first timers only issue, this issue is intended for someone who has never contributed to the project before so that we can help them to get started and find their feet.

@Ksaboor I think you're next in line. Would you like to take this one on?

**Comment by Ksaboor** [comment]
Yes I'll take this on. Thanks for this opportunity!

**Comment by wilkinsona** [comment]
Great! Thank you, @Ksaboor. If you have any questions, please don't hesitate to ask here.

**Comment by wilkinsona** [comment]
@ihrigb Yes, that's right. Thank you.

**Comment by Raheela1024** [comment]
Hi, can I work on it now?

**Comment by wilkinsona** [comment]
Thank you for the offer, @Raheela1024, but @Ksaboor is working on this one.

**Comment by Ksaboor** [comment]
@wilkinsona @philwebb 

Sorry for the delay on working this.

* Ok, I signed the Contributor License Agreement.
* fork, clone and created a separate branch for my work "Dev branch"
* But now I'm rereading the problem but I'm still having trouble try understanding it so I could provide a solution.

* Could I get some help understanding the issue on how the two blocks of code help me? 

From my understanding, I'm applying the TomcatConnectorCustomizer and TomcatContextCustomizer beans to the ServletWebServerFactoryConfiguration and ReactiveWebServerFactoryConfiguration.

Am I implementing the two blocks of code into ServletWebServerFactoryConfiguration class and the ReactiveWebServerFactoryConfiguration class?

**Comment by wilkinsona** [comment]
The two blocks of code above are an illustration of how the changes proposed here will improve things for users. In terms of implementing the proposed changes, you're right that they should be made in `ServletWebServerFactoryConfiguration` and `ReactiveWebServerFactoryConfiguration`. For example, the changes to apply connector customisers to the Servlet-based Tomcat factory will leave its bean method looking something like this:

```java
@Bean
public TomcatServletWebServerFactory tomcatServletWebServerFactory(
		ObjectProvider<TomcatConnectorCustomizer> connectorCustomizers) {
	TomcatServletWebServerFactory factory = new TomcatServletWebServerFactory();
	factory.getTomcatConnectorCustomizers().addAll(
			connectorCustomizers.orderedStream().collect(Collectors.toList()));
	return factory;
}
```

**Comment by tiwarivikash** [comment]
@Ksaboor Are you still working on it?

**Comment by Ksaboor** [comment]
Yes, I'm sorry I had to pause this week for interviews. I will continue working on it today.

**Comment by mbhave** [comment]
@Ksaboor Do you think you can give us an estimate on when you can submit a PR for this, please? If for some reason you aren't able to work on it by then, we can give someone else a chance.

**Comment by Ksaboor** [comment]
yeah sure, I'm working on it right now I should have a PR by Tuesday of this week. If not I'll happily give it away to the next person. :)

**Comment by Ksaboor** [comment]
Hey sorry but with the stress of job hunting and studying technical challenges. I'm going to have step away from this issue.

**Comment by tiwarivikash** [comment]
@mbhave Can I work on it?

**Comment by wilkinsona** [comment]
@Ksaboor Thanks for letting us know. Best of luck with your job hunting.

Thanks for the offer, @tiwarivikash, but, looking back through the comments I think @Raheela1024 is next in line. @Raheela1024 are you still interested in working on this?

**Comment by Raheela1024** [comment]
@wilkinsona Yes i will work on it. Thanks for this opportunity!

**Comment by Razi007** [comment]
@wilkinsona I guess i raised request before @Raheela1024.But no issue @Raheela1024  please continue.

**Comment by Raheela1024** [comment]
@wilkinsonaI  I have changed in `ServletWebServerFactoryConfiguration.java` and `ReactiveWebServerFactoryConfiguration.java` files. Below change is for `ServletWebServerFactoryConfiguration.java` but i need to know am in correct direction ?.
After changing that do I need to update test cases in `ServletWebServerFactoryAutoConfigurationTests.java` and `ReactiveWebServerFactoryAutoConfigurationTests.java` ?. There is any other area which I missed ?

 `ServletWebServerFactoryConfiguration.java` 

```
@Bean
public TomcatServletWebServerFactory tomcatServletWebServerFactory(
		ObjectProvider<TomcatConnectorCustomizer> connectorCustomizers,
		ObjectProvider<TomcatContextCustomizer> contextCustomizers) {
	TomcatServletWebServerFactory factory = new TomcatServletWebServerFactory();
	factory.getTomcatConnectorCustomizers().addAll(
			connectorCustomizers.orderedStream().collect(Collectors.toList()));
	factory.getTomcatContextCustomizers().addAll(
			contextCustomizers.orderedStream().collect(Collectors.toList()));
	return factory;
}

```

**Comment by wilkinsona** [comment]
@Raheela1024 Yes, that's exactly the right direction. Thank you.

It would be good to add 4 tests that verify that connector customiser and context customiser beans are added to the reactive and servlet factories. A test that verifies that a connector customiser bean is added to the servlet factory would like something like this:

```java
@Test
public void tomcatConnectorCustomizerBeanIsAddedToFactory() {
	WebApplicationContextRunner runner = new WebApplicationContextRunner(
			AnnotationConfigServletWebServerApplicationContext::new)
					.withConfiguration(AutoConfigurations
							.of(ServletWebServerFactoryAutoConfiguration.class))
					.withUserConfiguration(
							TomcatConnectorCustomizerConfiguration.class);
	runner.run((context) -> {
		TomcatServletWebServerFactory factory = context
				.getBean(TomcatServletWebServerFactory.class);
		assertThat(factory.getTomcatConnectorCustomizers()).hasSize(1);
	});
}

@Configuration
static class TomcatConnectorCustomizerConfiguration {

	@Bean
	public TomcatConnectorCustomizer connectorCustomizer() {
		return (connector) -> {
		};
	}

}
```

**Comment by Raheela1024** [comment]
@wilkinsona I have created the [PR ](https://github.com/spring-projects/spring-boot/pull/15492) Please review and let me know if missed something or doing wrongly.
